File: backend/app/schemas/ai.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.common import PaginationParams

AIProviderName = Literal["openai", "groq", "anthropic", "mock"]
ChatRole = Literal["system", "user", "assistant"]


class ChatMessage(BaseModel):
    role: ChatRole
    content: str = Field(min_length=1)


class ChatRequest(BaseModel):
    messages: list[ChatMessage] = Field(min_length=1)
    model: str | None = None
    stream: bool = False


class TokenUsage(BaseModel):
    prompt_tokens: int = Field(ge=0, default=0)
    completion_tokens: int = Field(ge=0, default=0)
    total_tokens: int = Field(ge=0, default=0)


class ChatResponse(BaseModel):
    content: str
    provider: AIProviderName
    model: str
    usage: TokenUsage


class StreamChunk(BaseModel):
    delta: str = ""
    done: bool = False


class AILogRead(BaseModel):
    id: str
    provider: AIProviderName
    model: str
    prompt_tokens: int = Field(ge=0)
    completion_tokens: int = Field(ge=0)
    latency_ms: int = Field(ge=0)
    created_at: datetime

    model_config = {"from_attributes": True}


class AILogListParams(PaginationParams):
    provider: AIProviderName | None = None

File: frontend/src/types/ai.ts
import type { ApiResponse, PaginatedResponse, PaginationParams } from "./api";

/** Mirror backend/app/schemas/ai.py */

export type AIProviderName = "openai" | "groq" | "anthropic" | "mock";
export type ChatRole = "system" | "user" | "assistant";

export interface ChatMessage {
  role: ChatRole;
  content: string;
}

export interface ChatRequest {
  messages: ChatMessage[];
  model?: string;
  stream?: boolean;
}

export interface TokenUsage {
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
}

export interface ChatResponse {
  content: string;
  provider: AIProviderName;
  model: string;
  usage: TokenUsage;
}

export interface StreamChunk {
  delta: string;
  done: boolean;
}

export interface AILogRead {
  id: string;
  provider: AIProviderName;
  model: string;
  prompt_tokens: number;
  completion_tokens: number;
  latency_ms: number;
  created_at: string;
}

export interface AILogListParams extends PaginationParams {
  provider?: AIProviderName;
}

export type ChatApiResponse = ApiResponse<ChatResponse>;
export type AILogListResponse = PaginatedResponse<AILogRead>;

/** WebSocket counter payload — mirror backend websocket contract */
export interface CounterMessage {
  count: number;
  timestamp: string;
}
